TP unique value error trying to store NULL IPv6 address for server

While updating MSO origin server properties in the new TP UI, I am getting 500 errors when trying to modify the second of the two servers. It appears that it is attempting to store a "null" IPv6 address for the server (server does not have IPv6 address).

```
[2017-11-13 20:41:18,833] [ERROR] DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: ERROR:  duplicate key value violates unique constraint "idx_140441_ip6_profile"
DETAIL:  Key (ip6_address, profile)=(, 192) already exists. [for Statement "UPDATE server SET ilo_username = ?, ip6_address = ? WHERE ( ( cachegroup = ? AND id = ? AND profile = ? AND status = ? AND type = ? ) )" with ParamValues: 1='None', 2='', 3='509', 4='2077', 5='192', 6='2', 7='3'] at /opt/traffic_ops/app/lib/API/Server.pm line 235
[2017-11-13 20:41:18,836] [ERROR]
```
We should not be attempting to store a new IPv6 address here, as it shouldn't have a value at all (and is not changing).

Problem does not seem to exist in the legacy Traffic Ops UI.

--- a/traffic_ops/app/lib/API/Server.pm
+++ b/traffic_ops/app/lib/API/Server.pm
@@ -200,36 +200,36 @@ sub update {
 	}
 
 	my $values = {
-		cachegroup       => $params->{cachegroupId},
-		cdn_id           => $params->{cdnId},
-		domain_name      => $params->{domainName},
-		host_name        => $params->{hostName},
-		https_port       => $params->{httpsPort},
-		ilo_ip_address   => $params->{iloIpAddress},
-		ilo_ip_netmask   => $params->{iloIpNetmask},
-		ilo_ip_gateway   => $params->{iloIpGateway},
-		ilo_username     => $params->{iloUsername},
-		ilo_password     => $params->{iloPassword},
-		interface_mtu    => $params->{interfaceMtu},
-		interface_name   => $params->{interfaceName},
-		ip6_address      => $params->{ip6Address},
-		ip6_gateway      => $params->{ip6Gateway},
-		ip_address       => $params->{ipAddress},
-		ip_netmask       => $params->{ipNetmask},
-		ip_gateway       => $params->{ipGateway},
-		mgmt_ip_address  => $params->{mgmtIpAddress},
-		mgmt_ip_netmask  => $params->{mgmtIpNetmask},
-		mgmt_ip_gateway  => $params->{mgmtIpGateway},
-		offline_reason   => $params->{offlineReason},
-		phys_location    => $params->{physLocationId},
-		profile          => $params->{profileId},
-		rack             => $params->{rack},
-		router_host_name => $params->{routerHostName},
-		router_port_name => $params->{routerPortName},
-		status           => $params->{statusId},
-		tcp_port         => $params->{tcpPort},
-		type             => $params->{typeId},
-		upd_pending      => $params->{updPending}
+		cachegroup			=> $params->{cachegroupId}, # required int
+		cdn_id				=> $params->{cdnId}, # required int
+		domain_name			=> $params->{domainName}, # required string
+		host_name			=> $params->{hostName}, # required string
+		https_port			=> ($params->{httpsPort}) ? $params->{httpsPort} : undef, # non zero int or null
+		ilo_ip_address		=> ($params->{iloIpAddress}) ? $params->{iloIpAddress} : undef, # non empty string or null
+		ilo_ip_netmask		=> ($params->{iloIpNetmask}) ? $params->{iloIpNetmask} : undef, # non empty string or null
+		ilo_ip_gateway		=> ($params->{iloIpGateway}) ? $params->{iloIpGateway} : undef, # non empty string or null
+		ilo_username		=> ($params->{iloUsername}) ? $params->{iloUsername} : undef, # non empty string or null
+		ilo_password		=> ($params->{iloPassword}) ? $params->{iloPassword} : undef, # non empty string or null
+		interface_mtu		=> $params->{interfaceMtu}, # required int
+		interface_name		=> $params->{interfaceName}, # required string
+		ip6_address			=> ($params->{ip6Address}) ? $params->{ip6Address} : undef, # non empty string or null
+		ip6_gateway			=> ($params->{ip6Gateway}) ? $params->{ip6Gateway} : undef, # non empty string or null
+		ip_address			=> $params->{ipAddress}, # required string
+		ip_netmask			=> $params->{ipNetmask}, # required string
+		ip_gateway			=> $params->{ipGateway}, # required string
+		mgmt_ip_address		=> ($params->{mgmtIpAddress}) ? $params->{mgmtIpAddress} : undef, # non empty string or null
+		mgmt_ip_netmask		=> ($params->{mgmtIpNetmask}) ? $params->{mgmtIpNetmask} : undef, # non empty string or null
+		mgmt_ip_gateway		=> ($params->{mgmtIpGateway}) ? $params->{mgmtIpGateway} : undef, # non empty string or null
+		offline_reason		=> ($params->{offlineReason}) ? $params->{offlineReason} : undef, # non empty string or null
+		phys_location		=> $params->{physLocationId}, # required int
+		profile				=> $params->{profileId}, # required int
+		rack				=> ($params->{rack}) ? $params->{rack} : undef, # non empty string or null
+		router_host_name	=> ($params->{routerHostName}) ? $params->{routerHostName} : undef, # non empty string or null
+		router_port_name	=> ($params->{routerPortName}) ? $params->{routerPortName} : undef, # non empty string or null
+		status				=> $params->{statusId}, # required int
+		tcp_port			=> ($params->{tcpPort}) ? $params->{tcpPort} : undef, # non zero int or null
+		type				=> $params->{typeId}, # required int
+		upd_pending			=> $params->{updPending} # required bool
 	};
 
 	my $rs = $server->update($values);
@@ -304,36 +304,36 @@ sub create {
 	}
 
 	my $values = {
-		cachegroup       => $params->{cachegroupId},
-		cdn_id           => $params->{cdnId},
-		domain_name      => $params->{domainName},
-		host_name        => $params->{hostName},
-		https_port       => $params->{httpsPort},
-		ilo_ip_address   => $params->{iloIpAddress},
-		ilo_ip_netmask   => $params->{iloIpNetmask},
-		ilo_ip_gateway   => $params->{iloIpGateway},
-		ilo_username     => $params->{iloUsername},
-		ilo_password     => $params->{iloPassword},
-		interface_mtu    => $params->{interfaceMtu},
-		interface_name   => $params->{interfaceName},
-		ip6_address      => $params->{ip6Address},
-		ip6_gateway      => $params->{ip6Gateway},
-		ip_address       => $params->{ipAddress},
-		ip_netmask       => $params->{ipNetmask},
-		ip_gateway       => $params->{ipGateway},
-		mgmt_ip_address  => $params->{mgmtIpAddress},
-		mgmt_ip_netmask  => $params->{mgmtIpNetmask},
-		mgmt_ip_gateway  => $params->{mgmtIpGateway},
-		offline_reason   => $params->{offlineReason},
-		phys_location    => $params->{physLocationId},
-		profile          => $params->{profileId},
-		rack             => $params->{rack},
-		router_host_name => $params->{routerHostName},
-		router_port_name => $params->{routerPortName},
-		status           => $params->{statusId},
-		tcp_port         => $params->{tcpPort},
-		type             => $params->{typeId},
-		upd_pending      => $params->{updPending}
+		cachegroup			=> $params->{cachegroupId}, # required int
+		cdn_id				=> $params->{cdnId}, # required int
+		domain_name			=> $params->{domainName}, # required string
+		host_name			=> $params->{hostName}, # required string
+		https_port			=> ($params->{httpsPort}) ? $params->{httpsPort} : undef, # non zero int or null
+		ilo_ip_address		=> ($params->{iloIpAddress}) ? $params->{iloIpAddress} : undef, # non empty string or null
+		ilo_ip_netmask		=> ($params->{iloIpNetmask}) ? $params->{iloIpNetmask} : undef, # non empty string or null
+		ilo_ip_gateway		=> ($params->{iloIpGateway}) ? $params->{iloIpGateway} : undef, # non empty string or null
+		ilo_username		=> ($params->{iloUsername}) ? $params->{iloUsername} : undef, # non empty string or null
+		ilo_password		=> ($params->{iloPassword}) ? $params->{iloPassword} : undef, # non empty string or null
+		interface_mtu		=> $params->{interfaceMtu}, # required int
+		interface_name		=> $params->{interfaceName}, # required string
+		ip6_address			=> ($params->{ip6Address}) ? $params->{ip6Address} : undef, # non empty string or null
+		ip6_gateway			=> ($params->{ip6Gateway}) ? $params->{ip6Gateway} : undef, # non empty string or null
+		ip_address			=> $params->{ipAddress}, # required string
+		ip_netmask			=> $params->{ipNetmask}, # required string
+		ip_gateway			=> $params->{ipGateway}, # required string
+		mgmt_ip_address		=> ($params->{mgmtIpAddress}) ? $params->{mgmtIpAddress} : undef, # non empty string or null
+		mgmt_ip_netmask		=> ($params->{mgmtIpNetmask}) ? $params->{mgmtIpNetmask} : undef, # non empty string or null
+		mgmt_ip_gateway		=> ($params->{mgmtIpGateway}) ? $params->{mgmtIpGateway} : undef, # non empty string or null
+		offline_reason		=> ($params->{offlineReason}) ? $params->{offlineReason} : undef, # non empty string or null
+		phys_location		=> $params->{physLocationId}, # required int
+		profile				=> $params->{profileId}, # required int
+		rack				=> ($params->{rack}) ? $params->{rack} : undef, # non empty string or null
+		router_host_name	=> ($params->{routerHostName}) ? $params->{routerHostName} : undef, # non empty string or null
+		router_port_name	=> ($params->{routerPortName}) ? $params->{routerPortName} : undef, # non empty string or null
+		status				=> $params->{statusId}, # required int
+		tcp_port			=> ($params->{tcpPort}) ? $params->{tcpPort} : undef, # non zero int or null
+		type				=> $params->{typeId}, # required int
+		upd_pending			=> $params->{updPending} # required bool
 	};
 
 	my $insert = $self->db->resultset('Server')->create($values);
@@ -1231,20 +1231,22 @@ sub is_server_valid {
 
 		# Validation checks to perform
 		checks => [
-			cachegroupId   => [ is_required("is required") ],
-			cdnId          => [ is_required("is required") ],
-			domainName     => [ is_required("is required") ],
-			hostName       => [ is_required("is required") ],
-			interfaceMtu   => [ is_required("is required") ],
-			interfaceName  => [ is_required("is required") ],
-			ipAddress      => [ is_required("is required") ],
-			ipNetmask      => [ is_required("is required") ],
-			ipGateway      => [ is_required("is required") ],
-			physLocationId => [ is_required("is required") ],
-			profileId      => [ is_required("is required") ],
-			statusId       => [ is_required("is required") ],
-			typeId         => [ is_required("is required") ],
-			updPending     => [ is_required("is required") ]
+			cachegroupId	=> [ is_required("is required"), is_like( qr/^\d+$/, "must be an integer" ) ],
+			cdnId			=> [ is_required("is required"), is_like( qr/^\d+$/, "must be an integer" ) ],
+			domainName		=> [ is_required("is required"), is_like( qr/^\S*$/, "must not contain spaces" ) ],
+			hostName		=> [ is_required("is required"), is_like( qr/^\S*$/, "must not contain spaces" ) ],
+			httpsPort		=> [ \&is_valid_port ],
+			interfaceMtu	=> [ is_required("is required"), is_like( qr/^\d+$/, "must be an integer" ) ],
+			interfaceName	=> [ is_required("is required") ],
+			ipAddress		=> [ is_required("is required") ],
+			ipNetmask		=> [ is_required("is required") ],
+			ipGateway		=> [ is_required("is required") ],
+			physLocationId	=> [ is_required("is required"), is_like( qr/^\d+$/, "must be an integer" ) ],
+			profileId		=> [ is_required("is required"), is_like( qr/^\d+$/, "must be an integer" ) ],
+			statusId		=> [ is_required("is required"), is_like( qr/^\d+$/, "must be an integer" ) ],
+			tcpPort			=> [ \&is_valid_port ],
+			typeId			=> [ is_required("is required"), is_like( qr/^\d+$/, "must be an integer" ) ],
+			updPending		=> [ is_required("is required") ]
 		]
 	};
 
@@ -1269,5 +1271,20 @@ sub is_valid_server_type {
 	}
 	return 0;
 }
+
+sub is_valid_port {
+	my ( $value, $params ) = @_;
+
+	if ( !defined $value or $value eq '' ) {
+		return undef;
+	}
+
+	if ( !( $value =~ /^\d+$/ ) ) {
+		return "invalid. Must be an integer.";
+	}
+
+	return undef;
+}
+
 
 1;

--- a/traffic_ops/app/t/api/1.2/server.t
+++ b/traffic_ops/app/t/api/1.2/server.t
@@ -143,6 +143,50 @@ ok $t->post_ok('/api/1.2/servers' => {Accept => 'application/json'} => json => {
 			"cdnId" => 100,
 			"domainName" => "example-domain.com",
 			"hostName" => "server2",
+			"httpsPort" => "string",
+			"interfaceMtu" => 1500,
+			"interfaceName" => "bond0",
+			"ipAddress" => "10.74.27.255",
+			"ipNetmask" => "255.255.255.252",
+			"ipGateway" => "10.74.27.194",
+			"physLocationId" => 100,
+			"profileId" => 100,
+			"statusId" => 1,
+			"typeId" => 1,
+			"updPending" => \0,
+		})
+		->status_is(400)
+		->json_is( "/alerts/0/level", "error" )
+		->json_is( "/alerts/0/text", "httpsPort invalid. Must be an integer." )
+	, "Does the server creation fail because httpsPort is a string?";
+
+ok $t->post_ok('/api/1.2/servers' => {Accept => 'application/json'} => json => {
+			"cachegroupId" => $cg2_mid_northwest,
+			"cdnId" => 100,
+			"domainName" => "example-domain.com",
+			"hostName" => "server2",
+			"interfaceMtu" => 1500,
+			"interfaceName" => "bond0",
+			"ipAddress" => "10.74.27.255",
+			"ipNetmask" => "255.255.255.252",
+			"ipGateway" => "10.74.27.194",
+			"physLocationId" => 100,
+			"profileId" => 100,
+			"statusId" => 1,
+			"tcpPort" => "string",
+			"typeId" => 1,
+			"updPending" => \0,
+		})
+		->status_is(400)
+		->json_is( "/alerts/0/level", "error" )
+		->json_is( "/alerts/0/text", "tcpPort invalid. Must be an integer." )
+	, "Does the server creation fail because tcpPort is a string?";
+
+ok $t->post_ok('/api/1.2/servers' => {Accept => 'application/json'} => json => {
+			"cachegroupId" => $cg2_mid_northwest,
+			"cdnId" => 100,
+			"domainName" => "example-domain.com",
+			"hostName" => "server2",
 			"interfaceMtu" => 1500,
 			"interfaceName" => "bond0",
 			"ipAddress" => "10.74.27.194",
@@ -298,30 +342,61 @@ ok $t->post_ok('/api/1.2/servers' => {Accept => 'application/json'} => json => {
 			"cdnId" => 100,
 			"domainName" => "northbound.com",
 			"hostName" => "tc2_ats2",
+			"httpsPort" => 443,
+			"iloIpAddress" => "",
+			"iloIpNetmask" => "",
+			"iloIpGateway" => "",
+			"iloUsername" => "",
+			"iloPassword" => "",
 			"interfaceMtu" => 1500,
 			"interfaceName" => "eth0",
+			"ip6Address" => "",
+			"ip6Gateway" => "",
 			"ipAddress" => "10.73.27.187",
 			"ipNetmask" => "255.255.255.0",
 			"ipGateway" => "10.73.27.1",
+			"mgmtIpAddress" => "",
+			"mgmtIpNetmask" => "",
+			"mgmtIpGateway" => "",
+			"offlineReason" => "",
 			"physLocationId" => 300,
 			"profileId" => 200,
+			"rack" => "",
+			"routerHostName" => "",
+			"routerPortName" => "",
+			"tcpPort" => 80,
 			"statusId" => 1,
 			"typeId" => 2,
 			"updPending" => \0,
 		})
     ->status_is(200)->or( sub { diag $t->tx->res->content->asset->{content}; } )
-    ->json_is( "/response/0/hostName" => "tc2_ats2")
-    ->json_is( "/response/0/domainName" => "northbound.com")
-    ->json_is( "/response/0/cachegroup" => "edge_atl_group")
-    ->json_is( "/response/0/ipNetmask" => "255.255.255.0")
-    ->json_is( "/response/0/interfaceName" => "eth0")
-    ->json_is( "/response/0/ipAddress" => "10.73.27.187")
-    ->json_is( "/response/0/ipGateway" => "10.73.27.1")
-    ->json_is( "/response/0/interfaceMtu" => "1500")
-    ->json_is( "/response/0/physLocation" => "HotAtlanta")
-    ->json_is( "/response/0/type" => "MID")
-    ->json_is( "/response/0/profile" => "MID1")
-            , 'Does the server details return?';
+		->json_is( "/response/0/cachegroup" => "edge_atl_group")
+		->json_is( "/response/0/domainName" => "northbound.com")
+		->json_is( "/response/0/hostName" => "tc2_ats2")
+		->json_is( "/response/0/httpsPort" =>443)
+		->json_is( "/response/0/iloIpAddress" => undef)
+		->json_is( "/response/0/iloIpNetmask" => undef)
+		->json_is( "/response/0/iloIpGateway" => undef)
+		->json_is( "/response/0/iloUsername" => undef)
+		->json_is( "/response/0/iloPassword" => undef)
+		->json_is( "/response/0/interfaceMtu" => "1500")
+		->json_is( "/response/0/interfaceName" => "eth0")
+		->json_is( "/response/0/ip6Address" => undef)
+		->json_is( "/response/0/ip6Gateway" => undef)
+		->json_is( "/response/0/ipNetmask" => "255.255.255.0")
+		->json_is( "/response/0/ipAddress" => "10.73.27.187")
+		->json_is( "/response/0/ipGateway" => "10.73.27.1")
+		->json_is( "/response/0/mgmtIpAddress" => undef)
+		->json_is( "/response/0/mgmtIpNetmask" => undef)
+		->json_is( "/response/0/mgmtIpGateway" => undef)
+		->json_is( "/response/0/offlineReason" => undef)
+		->json_is( "/response/0/physLocation" => "HotAtlanta")
+		->json_is( "/response/0/type" => "MID")
+		->json_is( "/response/0/profile" => "MID1")
+		->json_is( "/response/0/rack" => undef)
+		->json_is( "/response/0/routerHostName" => undef)
+		->json_is( "/response/0/routerPortName" => undef)
+	, 'Is the server created?';
 
 ok $t->get_ok('/api/1.2/servers/details')->status_is(400)->or( sub { diag $t->tx->res->content->asset->{content}; } ),
 	'Does the validation error occur?';
@@ -380,6 +455,55 @@ ok $t->put_ok('/api/1.2/servers/' . $svr_id  => {Accept => 'application/json'} =
     ->json_is( "/response/0/type" => "EDGE")
     ->json_is( "/response/0/profile" => "EDGE1")
             , 'Does the server details return?';
+
+ok $t->put_ok('/api/1.2/servers/' . $svr_id  => {Accept => 'application/json'} => json => {
+			"cachegroupId" => $edge_atl_group,
+			"cdnId" => 100,
+			"domainName" => "northbound.com",
+			"hostName" => "tc1_ats3",
+			"httpsPort" => 443,
+			"iloIpAddress" => "",
+			"iloIpNetmask" => "",
+			"iloIpGateway" => "",
+			"iloUsername" => "",
+			"iloPassword" => "",
+			"interfaceMtu" => 1500,
+			"interfaceName" => "eth0",
+			"ip6Address" => "",
+			"ip6Gateway" => "",
+			"ipAddress" => "10.74.27.186",
+			"ipNetmask" => "255.255.255.0",
+			"ipGateway" => "10.74.27.1",
+			"mgmtIpAddress" => "",
+			"mgmtIpNetmask" => "",
+			"mgmtIpGateway" => "",
+			"offlineReason" => "",
+			"physLocationId" => 100,
+			"profileId" => 100,
+			"rack" => "",
+			"routerHostName" => "",
+			"routerPortName" => "",
+			"tcpPort" => 80,
+			"statusId" => 1,
+			"typeId" => 1,
+			"updPending" => \0,
+		})
+    ->status_is(200)->or( sub { diag $t->tx->res->content->asset->{content}; } )
+    ->json_is( "/response/0/iloIpAddress" => undef)
+    ->json_is( "/response/0/iloIpNetmask" => undef)
+    ->json_is( "/response/0/iloIpGateway" => undef)
+    ->json_is( "/response/0/iloUsername" => undef)
+    ->json_is( "/response/0/iloPassword" => undef)
+    ->json_is( "/response/0/ip6Address" => undef)
+    ->json_is( "/response/0/ip6Gateway" => undef)
+    ->json_is( "/response/0/mgmtIpAddress" => undef)
+    ->json_is( "/response/0/mgmtIpNetmask" => undef)
+    ->json_is( "/response/0/mgmtIpGateway" => undef)
+    ->json_is( "/response/0/offlineReason" => undef)
+    ->json_is( "/response/0/rack" => undef)
+    ->json_is( "/response/0/routerHostName" => undef)
+    ->json_is( "/response/0/routerPortName" => undef)
+            , 'Are empty strings turned into null values?';
 
 ok $t->put_ok('/api/1.2/servers/' . $svr_id => {Accept => 'application/json'} => json => {
         "ipAddress" => "10.10.10.220",

